### PR TITLE
wallet: Deniability API (Unilateral Transaction Meta-Privacy)

### PR DESCRIPTION
**Pull request by denavila**
This PR is the wallet API and implementation portion of the GUI PR ( https://github.com/bitcoin-core/gui/pull/733 ) which is an implementation of the ideas in Paul Sztorc's blog post "**Deniability - Unilateral Transaction Meta-Privacy**"(https://www.truthcoin.info/blog/deniability/).

The GUI PR has all the details and screenshots of the GUI additions. Here I'll just copy the relevant context for the wallet API changes:

"_In short, Paul's idea is to periodically split coins and send them to yourself, making it look like common "spend" transactions, such that blockchain ownership analysis becomes more difficult, and thus improving the user's privacy.
We've implemented this as an additional "Deniability" wallet view. The majority of the code is in a new deniabilitydialog.cpp/h source files containing a new DeniabilityDialog class, hooked up to the WalletView class._ "

While the Deniability dialog can be implemented entirely with the existing API, adding the core "deniabilization" functions to the interfaces::Wallet API allows us to implement the GUI portion with much less code, and more importantly allows us to add RPC support and more thorough unit tests, which is what this PR attempts to do.

The main "deniabilization" functionality is implemented in spend.cpp/h and feebumper.cpp/h and exposed via the interfaces::Wallet API. 
We've added RPC support via a new RPC call "walletdeniabilizecoin". 
We also extended the Wallet tests with coverage for the functionality.

### CODE REVIEW

**Comment by denavila** [comment]
@achow101 - per our discussion, I've created a PR with the core "deniability" functionality in the main repo here. The GUI portion is rebased over this branch and is in the GUI repo. I've updated the PRs' descriptions to point to each other.
Please let me know what you think. Thanks!

**Review by ghost — CHANGES_REQUESTED** [review]
NACK

It doesn't improve privacy. Not considering fingerprints. A better way to improve privacy in bitcoin payments using bitcoin core could be payjoin. 

False sense of privacy implemented in this pull request only helps miners with fees.

**Comment by denavila** [comment]
> Bitcoin Core's wallet has a "unique" fingerprint on transactions that is shared by way less than 50% of transactions in the network ...

By fingerprint, do you mean Bitcoin Core's transactions can be identified on the block chain? If that's the case that's unfortunate for sure. What's the nature of the difference? Is it on purpose and if not, would it be possible to fix?

**Comment by denavila** [comment]
> See for example https://github.com/achow101/wallet-fingerprinting/blob/main/fingerprints.md

Thanks! 
At a quick glance, it looks like Core's transactions are somewhat close to Electrum's. 
I'm not familiar with some of the items, but assuming it's possible, I can look into inject some variety in the "deniabilized" transactions to make them resemble a variety of wallets (eg Electrum as a start). 
Is there prior work in this by any chance? That would help me for sure.
Also, should I pursue it for this PR, or would a follow up PR be acceptable as well?

**Comment by denavila** [comment]
> It doesn't improve privacy. 

Can you please elaborate why you think so?
Paul's blog post had some pretty compelling arguments why this scheme should work and I think it has a lot of merit.

> False sense of privacy implemented in this pull request only helps miners with fees.

It's important to point out that this PR doesn't add functionality that couldn't be achieved manually before.
If a user wanted to split a coin and send it to themselves, they could do it with the existing Bitcoin Core (either via RPC/Console or the GUI). It's just cumbersome and error prone.

The API we're adding in this PR makes this a simple and robust single operation (C++ API call or RPC call).
It's entirely opt-in and the user can decide to use it or not, up to their own judgement on how effective it is in improving their privacy. I would assume anyone using the C++/RPC interface understands the implications and can be trusted to know what they are doing.

The GUI PR (https://github.com/bitcoin-core/gui/pull/733) implements automation on top of this API, which is also entirely opt-in. We provide an explanation of how it works, so the user could make an educated decision on whether to use it or not. 
A lot of new Bitcoin users (myself including), weren't aware early on that it's important to use new addresses when receiving BTC. Many exchanges also don't make it easy either (eg by requiring time-locked whitelisting etc).
As a result I suspect many users may wish to improve their privacy by "fixing" their UTXOs after the fact.
This has been my main motivation to pursue creating this tool. I personally wanted to use it and I suspected others may appreciate it too.

**Comment by zkfrio** [comment]
> It's important to point out that this PR doesn't add functionality that couldn't be achieved manually before.
If a user wanted to split a coin and send it to themselves, they could do it with the existing Bitcoin Core (either via RPC/Console or the GUI). It's just cumbersome and error prone.

Good point and most of the people who tried this with coins involved in any incident regret doing it. Privacy in bitcoin is not as simple as doing some transactions with your own addresses automatically. There are lot of things you need to take care of. One mistake and your privacy is breached, publicly available for the whole world.

Payjoin will improve privacy without wasting fees on these automated transactions by hiding the amounts involved in transaction. Presently, payjoin is less used because of the need for a server but this can be improved.

**Comment by denavila** [comment]
> Good point and most of the people who tried this with coins involved in any incident regret doing it. Privacy in bitcoin is not as simple as doing some transactions with your own addresses automatically. There are lot of things you need to take care of. One mistake and your privacy is breached, publicly available for the whole world.

What kind of incidents do you mean? Do you have some concrete examples?
I don't see how one's privacy would be any worse after sending transactions to oneself. 
Each transaction moves further away from a potentially known UTXO, to new UTXOs that only the user's wallet knows are theirs. To a blockchain analysis that would look like a "spend" and they'll have to guess which one is potentially a different user and which one is a change address. With each split, the guesses become less and less accurate, thus leading to increased privacy, no?
If you see a flaw in this argument, please share you explanation, because I'm not seeing an issue.

**Review by ishaanam — COMMENTED** [review]
Hey @denavila, welcome to Bitcoin Core and thanks for working on this! While I think that this idea could be useful to some extent if implemented correctly, I have a few concerns about the current implementation:
- Currently there is no precaution against a user spending two outputs of a deniabilization transaction together. IMO I think that this is quite problematic and could lead to these deniabilization transactions being rendered ineffective. Not only that, but then it also becomes obvious that a user is trying to deniabilize their funds, which is something that should be avoided.
- As mentioned by @MarcoFalke, currently it is simple to identify Bitcoin Core transactions. Given that the Bitcoin Core wallet does not make up a large portion of on-chain transactions, if a blockchain observer sees that there is a long chain of transactions sending to and receiving from Bitcoin Core wallets, it could become clear that these are deniabilization transactions. 
- Additionally, in the current implementation all of the inputs of a deniabilization transaction must share the same address. It is not immediately obvious to me why this was done. This fact, combined with the previous two issues, could make it trivial to identify deniabilization transactions created by BItcoin Core. 

Solving some of these problems could be tricky and might require significantly more code being added to the Bitcoin Core wallet. Given this, I’m unsure whether it would make sense for an implementation of something like this to be built directly into Bitcoin Core. I'd be curious to hear what others think.

**Comment by zkfrio** [comment]
> > Good point and most of the people who tried this with coins involved in any incident regret doing it. Privacy in bitcoin is not as simple as doing some transactions with your own addresses automatically. There are lot of things you need to take care of. One mistake and your privacy is breached, publicly available for the whole world.
> 
> What kind of incidents do you mean? Do you have some concrete examples? I don't see how one's privacy would be any worse after sending transactions to oneself. 

Sorry. I cannot share exact incidents that involve "crime" on GitHub. However, you can look at this explanation from bitcoin privacy wiki to understand basics:

_"In a peeling chain, a single address begins with a relatively large amount of bitcoins. A smaller amount is then peeled off this larger amount, creating a transaction in which a small amount is transferred to one address, and the remainder is transferred to a one-time change address. This process is repeated - potentially for hundreds or thousands of hops - until the larger amount is pared down, at which point (in one usage) the amount remaining in the address might be aggregated with other such addresses to again yield a large amount in a single address, and the peeling process begins again[5]."_

https://en.bitcoin.it/wiki/Privacy

**Comment by denavila** [comment]
Hi @ishaanam  and thank you for looking at my PR.
I do agree the implementation can be improved and your feedback helps me greatly.

Here are my thoughts regarding your concerns:
1) I thought about this problem too and indeed spending deniabilized outputs is somewhat challenging. 
Obviously, if the user just HODLs, or spends less than a deniabilized UTXO, then all is good. 
However if they spend larger amounts, more than one UTXOs would get picked up and that would reveal that they belong to the same user. 
I don't think this destroys the entire privacy gained from deniabilization, since in a regular "spend" transaction with no-reuse tag, would look similar to deniabilized transaction, and blockchain analysis would've probably assigned some probability that some UTXOs are change, and our spending would confirm those ... But yes, the larger the spend the more would be revealed, so it's not ideal. 
Perhaps we could augment the UTXO selection algorithm to be deniabilization-aware, and perhaps there's ways to pick UTXO that minimize the damage? I'm not sure what heuristic we should use for that, so I'm open to suggestions.
Another option would be to provide a user parameter for "likely spendable size", which we use in the chunking decision and make strictly bigger chunks if possible. There's already logic that avoids "dust" chunks so it won't be difficult to extend in that direction.

2) Yes, the fingerprint concern is valid and unfortunate (and sadly I wasn't aware of it when I started this work). However it's also a pre-existing problem and while I do agree we should fix it, I'm hoping we could do this in a separate PR. Also I'm not yet knowledgable enough to tackle it, so if you guys want me to pursue a solution I could use some help.

3) The reason for grouping UTXOs that share the same address is that (and correct me if I'm wrong) if we can spend one UTXO from a given address, then we have ownership of all UTXOs that share that address. In other words, we don't lose any privacy by deniabilizing the whole group. 
In my limited experience, large amount of shared-address UTXOs tends to happen when transferring from an exchange (eg periodic buys) - often people don't update their withdrawal address, and many exchanges make it difficult to update (eg by requiring whitelisting and so on).
This grouping allows the deniabilization process to actually reduce the UTXO count if used for just few cycles.
At any rate, it would be trivial do disable the grouping if there's a flow in this reasoning, or otherwise have it on a user toggle or some such.

> Given this, I’m unsure whether it would make sense for an implementation of something like this to be built directly into 
Bitcoin Core

I thought about if this could be implemented in some kind of a plugin, but as far as I know, there's no support for plugins yet?
Originally my PR was entirely in the GUI, since I wanted to avoid API changes, but @achow101 correctly pointed out it would be safer and easier to test the main functions as part of the wallet API, so that's why I split the PR into a core and GUI parts.
If there's a path to implement this in a more "plugin" way, I'd be interested to hear of course.

Thank you for your feedback. I greatly appreciate it.

**Comment by murchandamus** [comment]
> 1. But yes, the larger the spend the more would be revealed, so it's not ideal.

Not only that, but if you spend a UTXO that’s the change from the previous transaction with the output of the next, that’s one thing. But if you spend an output with the change from fifteen transactions prior, you may heavily imply that all payments in between were also self-sends.

> * Another option would be to provide a user parameter for "likely spendable size", which we use in the chunking decision and make strictly bigger chunks if possible. There's already logic that avoids "dust" chunks so it won't be difficult to extend in that direction.

Yeah, that sounds like a reasonable direction of development. It would e.g. be possible to pick from two otherwise similar inputs sets the one that creates a change output whose value is most different from other UTXOs already in the wallet. Otherwise, it is a bit difficult to predict what a “likely spendable size” is, since it depends on many different factors such as future exchange rate, and unknown future motivations for the user to create transactions. I would surmise that past transactions would only have a limited utility in predicting future spending sizes. In fact in my master-thesis, I explored a similar idea by attempting to always create change of a similar amount as the recipient output. Unfortunately, the resulting UTXO pool performed subpar.

**Comment by kravets** [comment]
In a tx graph ( tree really with large and randomized fan out, eventual merging of some UTXOs does not necessarily indicate that the later clustered entity is the same entity KYC’s on an exchange prior to the deniabilization fan out

**Comment by kravets** [comment]
Would it not be better to postpone removing the Core tx signatures to Deniabilization 1.1 or 2.0 version and get this safe, isolated UI centric feature merged in sooner rather than later ?

**Comment by denavila** [comment]
> Needs rebase if still relevant

Ah, I didn't notice, sorry. I'll rebase it now.

**Comment by denavila** [comment]
I hit an issue in my wallet tests after rebasing. Looks like a bug was introduced in ExtractDestination. I left a comment in https://github.com/bitcoin/bitcoin/pull/28246

File: src/wallet/spend.cpp [hunk 1]
```diff
@@ -1392,4 +1392,302 @@ bool FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& nFeeRet,
...
+    for (const auto& coin : preset_inputs.coins) {
+        CTxDestination destination = CNoDestination();
+        ExtractDestination(coin->txout.scriptPubKey, destination);
+        if (!std::get_if<CNoDestination>(&destination) && !op_shared_destination) {
+            op_shared_destination = destination;
+        }
+        if (!op_shared_destination || !(*op_shared_destination == destination)) {
+            return util::Error{_("Input addresses must all match.")};
+        }
+    }
```
Comment (denavila) [inline src/wallet/spend.cpp:1597]: @achow101 - continuing our PubKeyDestination discussion here.
In the highlighted code I was trying to determine if all inputs share the same address or not. 
This seemed to work in all my tests, for real wallets and for unit test wallets (eg from coinbase keys), however with the introduction of the PubKeyDestination address, this now breaks for coinbase keys. In particular the call to OutputTypeFromDestination further down in this function, now returns nullopt.
I suspect my understanding of scripts and addresses is flawed. I assumed scripts is just the general encoding of addresses on the block chain, while addresses (CTxDestination) are specialized classes of address types that Bitcoin Core understands, and I assumed CNoDestination is just addresses we don't understand and can't use for transactions. I'm guessing there's more to it?
How should I modify my code to correctly handle the new PubKeyDestination, which seems to be how coinbase key wallet scripts get classified now?
I added a quick hack to pass the CI below, but it sounds like this is probably not correct behavior (to treat PubKeyDestination as OutputType::Legacy).
Thank you in advance!

Comment (achow101) [reply src/wallet/spend.cpp:1597]: I suggest that you operate entirely on `CScript`s rather than `CTxDestination`s. Under the previous semantics, P2PK and P2PKH scripts would be bundled together which I don't think is actually the behavior that you want. It would be better to use `CScript` and the `Solver` to get precise script types, and avoid using `CTxDestination` and `OutputType`. The latter two are more for user input and display rather than providing precise distinction between scripts.

Comment (denavila) [reply src/wallet/spend.cpp:1597]: How do I compare one Script with another (in particular one txout.scriptPubKey with another). 
The idea being that I want to issue a transaction where inputs are strictly from inputs that share the same "address" (address in layman terms - the address that people copy/paste to receive BTC).

Comment (denavila) [reply src/wallet/spend.cpp]: I think I figured it out. CScript appears to have an operator== so I switched to using that. I'm hoping this works in the way I wanted to compare equality of laymen "addresses".
 
I pushed an update to the PR. Please let me know if this makes sense.

File: src/wallet/spend.cpp [hunk 2]
```diff
@@ -1392,4 +1392,293 @@ bool FundTransaction(CWallet& wallet, CMutableTransaction& tx, CAmount& nFeeRet,
...
+        return util::Error{strprintf(_("Insufficient amount (%d) for a deniabilization transaction, min amount (%d), tx fee (%d)."), total_amount, min_post_split_amount, estimated_tx_fee)};
+    }
+
+    CTxDestination shared_destination = CNoDestination();
+    ExtractDestination(shared_script, shared_destination);
+    std::optional<OutputType> op_output_type = OutputTypeFromDestination(shared_destination);
+    if (!op_output_type) {
+        op_output_type = wallet.m_default_change_type;
+    }
+    OutputType output_type = op_output_type ? *op_output_type : wallet.m_default_address_type;
```
Comment (denavila) [inline src/wallet/spend.cpp:1618]: @achow101 m
Is this a correct way to get an OutputType from a CScript? I need the OutputType for the ReserveDestination constructor.
For coinbase key wallet scripts, the destination gets classified as PubKeyDestination, thus OutputTypeFromDestination will return nullopt, so we'll end up using the wallet.m_default_address_type.
This appears to work, but I'm not sure if it's kosher ...

Comment (achow101) [reply src/wallet/spend.cpp:1618]: That looks fine.

**Comment by denavila** [comment]
@maflcko - The CI has been failing on the "fuzzer" test with a "time out", and I re-tried it few times, but to no avail. I'm not sure if this is a global CI issue, or if it is indeed something wrong with my code. Is there a chat/forum I can inquire about CI issues?

**Comment by 1440000bytes** [comment]
Concept ACK

I cant comment on the approach before testing and going through everything. I had tried to implement similar feature in electrum: https://github.com/spesmilo/electrum/pull/8159

Fingerprinting issue could be resolved by using random fingerprints for transactions in this feature or add spoofing feature in core: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2023-October/021998.html

BTCPay Server also randomizes wallet fingerprint for transactions: https://github.com/btcpayserver/btcpayserver/blob/a921504bcf619c5e845813b8f994b39147694a97/Changelog.md?plain=1#L2079

**Comment by denavila** [comment]
@1440000bytes , @maflcko, @achow101  
In commit [c6e1d30 ](https://github.com/bitcoin/bitcoin/commit/c6e1d30fee58f716f1cb7d6c31fc1a54c1cc4d89) I implemented a first pass "fingerprint spoofing" for deniabilization transactions, based on the pointers you guys gave me. Please let me know what you think. 
Thanks!

**Comment by ishaanam** [comment]
@denavila 
> If there's a path to implement this in a more "plugin" way, I'd be interested to hear of course.

Technically this could be written as a python script that uses Bitcoin Core RPCs to create and sign transactions, but does the deniabilization steps and anti-fingerprinting measures in python. 

That being said, I also have a few reservations about the fingerprint imitation implemented here:
- It seems like it introduces a big maintenance burden because whenever another wallet changes the fingerprints they exhibit, this will need to updated correspondingly.
- Additionally, I'm not sure how much this actually fixes. Because this does not eliminate all of Bitcoin Core's fingerprints (which would be quite a challenge), combining another wallet's fingerprints with Bitcoin Core's could just produce a more obvious fingerprint which will make it obvious that a given transaction is a deniability transaction. 
- Lastly, some of the fingerprints that are removed when certain wallets are imitated (like `standardVersion` and `antiFeeSniping`) are actually useful. `antiFeeSniping` is something that we want to be doing because it prevents fee-sniping. In fact, we should be encouraging other wallets to do this instead of removing it from our own transactions. Additionally, the `nVersion` of a transaction may need to have a certain value depending on which time locks are being used (this is not implemented yet, but now that Bitcoin Core has miniscript, they likely will be soon).

File: src/wallet/spend.cpp [hunk 3]
```diff
@@ -1557,6 +1557,69 @@ std::pair<unsigned int, bool> CalculateDeniabilizationCycles(CWallet& wallet, co
...
+    // https://github.com/achow101/wallet-fingerprinting/blob/main/fingerprints.md
+    // https://gitlab.com/1440000bytes/goldfish
+    // https://ishaana.com/blog/wallet_fingerprinting/
+    static const Fingerprint s_walletFingerprints[] = {
+        {true, true, false},   // Core
+        {true, true, true},    // Electrum
+        {true, false, false},  // Blue
+        {false, false, true},  // Trezor
+        {false, false, false}, // Trust, Ledger
+    };
```
Comment (1440000bytes) [inline src/wallet/spend.cpp]: Can also [randomize RBF](https://github.com/btcpayserver/btcpayserver/pull/1596) to make fingerprinting more difficult.

Comment (denavila) [reply src/wallet/spend.cpp:1579]: I was thinking about randomizing RBF - however we issue the deniabilization transactions with the lowest fee possible (eg confirm_target = 100+ for daily frequency, 1000+ for weekly frequency), so there's a chance the TX may get stuck if the mempool suddenly gets crowded. If that happens we attempt an RBF at said frequency, so I think it's important to have it always enabled.
Alternatively we could randomize RBF, and use a conservative fee calculation in the RBF=off case, but that would likely be more expensive ...

**Comment by denavila** [comment]
@ishaanam - Thank you for your feedback!

Yes, it's true Deniability can be implemented without changing the API. Indeed, my first implementation was entirely in the GUI source code. Probably it can also be done in Python and RPC, but Python is not my forte so I wouldn't attempt that myself.

Having said that, @achow101 suggested that we move the core deniability logic in the Core source code (eg in the CWallet implementation), as that would make it more robust, easier to test and allow exposing it to RPC. I agreed with this assessment, and indeed after splitting off the core functionality, the GUI portion became much more straightforward.
At any rate, if the API or GUI PRs don't get merged, implementing this via RPC / script would be the way to go.

Regarding the fingerprinting:
1) The maintenance concern is legitimate, and while not super time sensitive (since users may use old wallets for a long time), it does mean we'd need to revisit this form time to time. Let me know if you have ideas how we can do this in a more future proof way? 
2) The implementation I put together above was just a first stab to get feedback. I implemented the 3 fingerprints that were easy and I understood. I'd be happy to implement more with some help. For example what are positive/negative EV inputs?
3) Currently I'm just using the fingerprint spoofing for deniabilization transactions, which by design use the lowest fees, and so hopefully they won't encourage fee sniping even if we use the nLockTime = 0. But I do agree, for the general case, we do want it, so maybe I can add an additional control to randomize in favor of sniping more often ... Let me know if you have ideas how we can improve on that.
I can also add nVersion dependence on the time lock. What should I check for?

File: src/wallet/spend.cpp (hunk 3 above)
Comment (1440000bytes) [reply src/wallet/spend.cpp]: - Transactions can be replaced even if they do not signal RBF since some nodes and miners support full RBF
- I don't think using core fee estimation for these transactions is a good idea

Comment (denavila) [reply src/wallet/spend.cpp:1579]: Ah, interesting, yeah if we can rely on full-RBF nodes, then we can randomize the flag.
Is Core's fee estimation also fingerprint-able? I could randomize that too, but what is a good alternative estimation?

Comment (denavila) [reply src/wallet/spend.cpp]: I added RBF spoofing in the latest commit.

Comment (denavila) [reply src/wallet/spend.cpp]: Hmm, so it doesn't look like feebumper::CommitTransaction would let me bump a transaction if the RBF flag wasn't set on the original transaction. For now I worked around it by passing an extra flag into SpoofTransactionFingerprint to only signal no-rbf when RBF is not required (as is the case with deniabilizaiton transactions). Let me know if you have better ideas how to work around this and take advantage of full-RBF nodes bumping regardless of the flag ...

**Comment by 1440000bytes** [comment]
@denavila  I am okay with this merged for testnet or signet alleast

We want to improve privacy and see results

Thats not how things work here

**Comment by 1440000bytes** [comment]
> @denavila I am okay with this merged for testnet or signet alleast
> 
> We want to improve privacy and see results
> 
> Thats not how things work here

Please join to @luke-jr in https://github.com/bitcoinknots/bitcoin

**Comment by hebasto** [comment]
A failure in the "Win64 native, VS 2022" CI job is unrelated. See: https://github.com/bitcoin/bitcoin/pull/28905.

**Comment by aureleoules** [comment]
This pull silently conflicts with master:
```make
wallet/rpc/spend.cpp: In lambda function:
wallet/rpc/spend.cpp:1806:49: error: ‘transaction_identifier<has_witness>::transaction_identifier(const uint256&) [with bool has_witness = false]’ is private within this context
 1806 |                 COutPoint outpoint(txid, nOutput);
      |                                                 ^
In file included from ./primitives/transaction.h:14,
                 from ./consensus/validation.h:11,
                 from wallet/rpc/spend.cpp:5:
./util/transaction_identifier.h:16:5: note: declared private here
   16 |     transaction_identifier(const uint256& wrapped) : m_wrapped{wrapped} {}
      |     ^~~~~~~~~~~~~~~~~~~~~~
```

**Comment by 1440000bytes** [comment]
Tested it and I think it can be improved further before users can try it on mainnet: https://github.com/bitcoin-core/gui/pull/733#issuecomment-2081539227

**Comment by saadbitcoin** [comment]
Can you healp me

في الأحد، ٢٨ أبريل ٢٠٢٤ ٧:٢٦ م 1440000bytes ***@***.***> كتب:

**Comment by achow101** [comment]
This PR does not seem to have attracted much attention from reviewers. As such, it does not seem important enough right now to keep it sitting idle in the list of open PRs.

Closing due to lack of interest.

**Comment by 1440000bytes** [comment]
Its true there was a lack of interest to review this pull request and its not the best tool for privacy. However, a worse way to do the same thing was used in Samourai earlier and still available in Ashigaru. Ricochet fee is 0.001 BTC for the default 4 hops, with an additional fee for each extra hop beyond 4.

image


http://ashicodepbnpvslzsl2bz7l2pwrjvajgumgac423pp3y2deprbnzz7id.onion/Ashigaru/Ashigaru-Mobile/src/commit/6b37843e5d1f7748836471cd1e3621a5fb1cc29f/app/src/main/java/com/samourai/wallet/ricochet/RicochetMeta.java#L65

Users have paid more than [2.8 BTC in fees](https://blockstream.info/address/bc1q52fzmcdqu07j845q7jnwzw9q68s924wdva3lfz) for ricochet service in past and it was even mentioned in the [DOJ press release](https://www.justice.gov/usao-sdny/pr/founders-and-ceo-cryptocurrency-mixing-service-arrested-and-charged-money-laundering#:~:text=ricochet).

So, it would have been great if users had a better alternative which has no service fee.

**Comment by 1440000bytes** [comment]
An alternative for anyone reading this pull request in the future: https://github.com/kristapsk/bitcoin-scripts/blob/master/ricochet-send.sh

It's a bash script that uses bitcoin core wallet.
